### ISSUE TITLE
Add Dispatcher.Created static event

### ISSUE THREAD

**Issue by matthew-a-thomas** [dotnet/wpf#7301]
# The need

It's undesirable to instantiate a [`Dispatcher`](https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/Threading/Dispatcher.cs) on a non-UI thread. Doing so is (usually) a bug and leads to several issues, such as:

* Deadlock. Usually the dispatcher will not be running. But code may check if the current thread has a dispatcher and if so post a callback to it and wait to proceed until the callback has executed. Since the dispatcher is not running the callback never executes
* Memory leak. Usually the dispatcher will not be running. But code may post callbacks to it. Those callbacks contain state which is strongly-referenced by the dispatcher. That state graph will not be garbage collected until the thread ends

However it is extremely easy to instantiate a `Dispatcher` on a non-UI thread. Broadly speaking there are two general ways this tends to happen:

* Access `Dispatcher.CurrentDispatcher`
* Instantiate anything that derives from `DispatcherObject`

It is usually possible for the disciplined solo programmer to avoid these issues in a small project. But sometimes people forget, and other times these issues are subtle and unexpected (e.g. I just recently learned that simply calling `CommandManager.InvalidateRequerySuggested()` creates a dispatcher). So these issues tend to crop up in larger projects with many developers. As a result it's virtually guaranteed that the software project will periodically encounter deadlocks, memory leaks, and other unexpected behavior.

# A solution

I'd like to solve this problem by giving developers a proactive way to know when a dispatcher is instantiated. Developers can write `#if DEBUG` that subscribes to the proposed event and inspects `Thread.CurrentThread` in the handler, and then log the stack trace when that happens. Then developers will know _exactly_ where the offending code is and will be able to solve it promptly.

# Alternatives

I think this would be better than the alternatives. Currently I only know of three alternatives. In increasing order of hackiness:

1. Use the Win32 API to be notified after a Win32 window is created. I assume such a hook would execute asynchronously, so this would really be no better than # 2, which is...
1. Use reflection to poll `Dispatcher._dispatchers` (while under the `Dispatcher._globalLock` guard of course)
1. Add a hook to intercept the call to the user32.dll > CreateWindowEx Win32 function call that happens while instantiating a dispatcher
1. Edit the CLR method descriptor for `Dispatcher.ctor` at runtime to point to my own hacked up constructor method

# The proposal

```csharp
public sealed class Dispatcher
{
  public static event DispatcherCreatedEventHandler? Created; // Add this event

  public static Dispatcher CurrentDispatcher
  {
    get
    {
      Dispatcher currentDispatcher = FromThread(Thread.CurrentThread);;
      if(currentDispatcher == null)
      {
        currentDispatcher = new Dispatcher();

        //// The following line is new ////
        Created?.Invoke();
        //// The above line is new ////

      }
      return currentDispatcher;
    }
  }
}

// Add this delegate
public delegate void DispatcherCreatedEventHander();
```

**Comment by miloush** [comment]
I do want to point out that using dispatcher on non-UI threads for dispatching prioritized operations (unrelated to UI) is a working scenario that I have used in the past.

I do not have anything against the API in principle (apart perhaps it could be just `EventHandler`), I am not so convinced by the motivation though, which seems to be to diagnose bugs in misbehaving code. (And therefore for example the event counterpart of dispatcher shutting down is not considered.) If that is the only need for the API, I think broader support for the proposal would help.

The `Dispatcher` utilizes several ETW events though. I am wondering whether introducing a dispatcher created ETW event instead would address your scenario?

**Comment by lindexi** [comment]
I think multi-threaded programming requires developers to have programming knowledge. This means that multi-threaded programming is not suitable for providing simple APIs, otherwise novice developers will accidentally use them and cause problems.

Keeping it complex can force developers to learn more about multi-threaded programming.

**Comment by matthew-a-thomas** [comment]
@miloush

> I do want to point out that using dispatcher on non-UI threads for dispatching prioritized operations (unrelated to UI) is a working scenario that I have used in the past.

That's definitely valid. It's situations like that that kept me from proposing something more... drastic. Such as "block dispatcher creation on XYZ threads".

> I do not have anything against the API in principle (apart perhaps it could be just `EventHandler`), I am not so convinced by the motivation though, which seems to be to diagnose bugs in misbehaving code. (And therefore for example the event counterpart of dispatcher shutting down is not considered.) If that is the only need for the API, I think broader support for the proposal would help.

Regarding `EventHandler`, I considered that but then we'd have to find something to use for both the `object? sender` and the `EventArgs e` parameters, and nothing comes to my mind that isn't readily available to subscribers&mdash;they can get the current dispatcher and the current thread, and other than that I don't see any other candidates.

Instead I tend to go the other direction. In fact I think I like just plain `Action`.

Regarding the motivation, yes I recognize there isn't much steam behind this proposal. But that's all the steam I have 😉

> The `Dispatcher` utilizes several ETW events though. I am wondering whether introducing a dispatcher created ETW event instead would address your scenario?

I'm unfamiliar with ETW events. Is that [Event Tracing for Windows](https://learn.microsoft.com/en-us/windows/win32/etw/about-event-tracing)?

**Comment by matthew-a-thomas** [comment]
@lindexi

> I think multi-threaded programming requires developers to have programming knowledge. This means that multi-threaded programming is not suitable for providing simple APIs, otherwise novice developers will accidentally use them and cause problems.
> 
> Keeping it complex can force developers to learn more about multi-threaded programming.

I'm not sure I'm following your train of thought. Are you suggesting a more complicated solution to what I described in "The need"?

**Comment by Laniusexcubitor** [comment]
The fact that CurrentDispatcher is creating a new not running dispatcher on any thread that does not have one yet does bite you when using WeakEventManager, and it will create memory leaks, that people want to actually avoid by using the construct.
The WeakEventTable will post it's cleanup code to the dispatcher, which creates a memory leak mentioned in the first post.

**Comment by lindexi** [comment]
@Laniusexcubitor I'd love to hear your thoughts.

**Comment by Laniusexcubitor** [comment]
Actually I would have preferred that we would not be in this state.
I think `Dispatcher.CurrentDispatcher` should have never just added a `Dispatcher`, and code that depends on a `Dispatcher` like `WeakEventManager` should have thrown exceptions, when called from a non dispatcher thread. 
For advanced scenarios, where you want to have a dispatcher on an additional UI thread you would just create a new dispatcher (perhaps have a `Dispatcher.Create()` method), after all in all those scenarios you need to call `Dispatcher.Run()` for the scenario to work.

But now there are applications out there that are build with the current behavior. Probably some of them creating small memory leaks because dispatchers get added to a threadpool thread.
Many constructs in WPF currently do create new dispatchers unintentionally without people realizing.
I was quite amazed when I realized some years ago, that WeakEventManager is working like it does. Of course it is a UI construct. But when looking at the API surface, it seems that you can use it independently of the UI thread.

I think ideally we would get a warning/error just like with BindingErrors (e.g. using a new PresentationTraceSource). 
That way there is at least a warning when debugging and people are made aware of potential problems. And for people who also want to log such errors in production, they could just add a listener to the PresentationTraceSource.

**Comment by lindexi** [comment]
Warnings aren't something devs pay attention to, so their usefulness feels pretty limited tbh. As for WPF, it does support multithreading — I'm concerned that throwing exceptions could end up being overly destructive.

**Comment by Laniusexcubitor** [comment]
Yes, that's why I would not go with the exception route. However I think attaching a `Dispatcher` without calling `Dispatcher.Run()` is always an error. So when providing a method for people to Create a new dispatcher for the current thread, showing actual intent to do so, and throwing an exception otherwise could also be a valid way. This would be a breaking change, but for people intentionally creating a new dispatcher it would be easily fixable by using the new creation method.
For people with current unintentional dispatcher creations, it is probably a valid notification of problems in their code, but there probably also should be a compatibility switch in that case, to allow them a YOLO-ignore of those problems.

It is true, that people often don't pay attention to warnings in the output window. But at least it is a way for people who do, to be informed about this dispatcher problem. But with the exceptions, as I mentioned there would also be the need to have a compatibility-switch. People having the problem WILL say: "My application worked perfectly fine, why now exceptions?". Memory leaks were potentially low, and who cares about creeping memory leaks, if it is a desktop application that runs a few hours, perhaps days and it does not consume all the memory? So those people would ignore the warnings. People perhaps looking to find a memory leak would probably also take a look at warnings.

**Comment by lindexi** [comment]
After hearing your take, I'm coming around to the design proposed in the issue — adding a static event would really help in this scenario.

**Comment by Laniusexcubitor** [comment]
Yes, the event would help in finding code that adds dead dispatchers. I would mark the event with `[System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]` (just like [CheckAccess](https://github.com/dotnet/wpf/blob/a0553fd3fbc5fd74f70ab23f36e8b62524739d80/src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/Threading/Dispatcher.cs#L193))
It is an event not intended for normal use. However I never understood the usage of `Never` on `CheckAccess` 😉
